File: src/derived-data.ts
import { promises as fs, type Dirent } from 'node:fs';
import os from 'node:os';
import path from 'node:path';
import { execa } from 'execa';

const DEFAULT_ROOT = path.join(os.homedir(), 'Library', 'Developer', 'Xcode', 'DerivedData');

export interface DerivedDataEntry {
  /** Absolute path of the folder, e.g. `.../DerivedData/MyApp-abcdef…`. */
  path: string;
  /** Folder name — the same string Xcode shows in its build locations. */
  name: string;
  /** Project or workspace this folder was built for, from `info.plist`. */
  workspacePath: string;
}

export interface SizedDerivedData extends DerivedDataEntry {
  /** Disk usage in bytes. */
  size: number;
  sizeLabel: string;
}

/**
 * Where Xcode keeps DerivedData. Honours a custom *absolute* location from Xcode's
 * preferences; a relative one ("Relative to Workspace") lives inside the project and
 * is already removed along with the worktree, so the default root still applies.
 * `LCC_DERIVED_DATA` overrides both.
 */
export async function derivedDataRoot(): Promise<string> {
  const override = process.env.LCC_DERIVED_DATA?.trim();
  if (override) return path.resolve(override.replace(/^~(?=$|\/)/, os.homedir()));
  try {
    const { stdout } = await execa('defaults', [
      'read',
      'com.apple.dt.Xcode',
      'IDECustomDerivedDataLocation',
    ]);
    const custom = stdout.trim();
    if (custom && path.isAbsolute(custom)) return custom;
  } catch {
    // Key unset — Xcode is using the default location.
  }
  return DEFAULT_ROOT;
}

/**
 * Every per-project folder under `root`. Xcode's own shared caches
 * (`ModuleCache.noindex`, `SDKStatCaches.noindex`, …) carry no `info.plist` and are
 * skipped — they belong to no project and must never be treated as removable.
 */
export async function listDerivedData(root: string): Promise<DerivedDataEntry[]> {
  let dirents: Dirent[];
  try {
    dirents = await fs.readdir(root, { withFileTypes: true });
  } catch {
    return [];
  }
  const entries: DerivedDataEntry[] = [];
  for (const dirent of dirents) {
    if (!dirent.isDirectory() || dirent.name.startsWith('.')) continue;
    const full = path.join(root, dirent.name);
    const workspacePath = await readWorkspacePath(full);
    if (!workspacePath) continue;
    entries.push({ path: full, name: dirent.name, workspacePath });
  }
  return entries;
}

async function readWorkspacePath(dir: string): Promise<string | null> {
  const plist = path.join(dir, 'info.plist');
  let raw: string;
  try {
    raw = await fs.readFile(plist, 'utf8');
  } catch {
    return null;
  }
  const direct = matchWorkspacePath(raw);
  if (direct) return direct;
  // Some Xcode versions write a binary plist — convert before matching.
  try {
    const { stdout } = await execa('plutil', ['-convert', 'xml1', '-o', '-', plist]);
    return matchWorkspacePath(stdout);
  } catch {
    return null;
  }
}

function matchWorkspacePath(xml: string): string | null {
  const match = /<key>WorkspacePath<\/key>\s*<string>([^<]*)<\/string>/.exec(xml);
  const value = match?.[1]?.trim();
  return value ? decodeEntities(value) : null;
}

function decodeEntities(value: string): string {
  return value
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&apos;/g, "'")
    .replace(/&amp;/g, '&');
}

/**
 * Folders whose project lives inside `worktreePath`. A worktree can own more than
 * one (an `.xcodeproj` and a `Package.swift`, say), so all matches come back.
 */
export async function forWorktree(
  entries: DerivedDataEntry[],
  worktreePath: string,
): Promise<DerivedDataEntry[]> {
  // Xcode records the resolved path; `git worktree list` does not resolve symlinks.
  const root = await realpath(worktreePath);
  return entries.filter((entry) => isInside(root, entry.workspacePath));
}

/** Entries whose project is gone from disk — the worktree was removed long ago. */
export async function orphans(entries: DerivedDataEntry[]): Promise<DerivedDataEntry[]> {
  const checked = await Promise.all(
    entries.map(async (entry) => ((await exists(entry.workspacePath)) ? null : entry)),
  );
  return checked.filter((entry): entry is DerivedDataEntry => entry !== null);
}

/** Attach disk usage to each entry, measuring a few folders at a time. */
export async function withSizes(entries: DerivedDataEntry[]): Promise<SizedDerivedData[]> {
  const sized: SizedDerivedData[] = new Array(entries.length);
  let next = 0;
  const worker = async () => {
    for (;;) {
      const index = next++;
      const entry = entries[index];
      if (!entry) return;
      const size = await dirSize(entry.path);
      sized[index] = { ...entry, size, sizeLabel: formatBytes(size) };
    }
  };
  await Promise.all(Array.from({ length: Math.min(8, entries.length) }, worker));
  return sized;
}

/** Disk usage in bytes via `du` — walking millions of build artifacts in JS is far slower. */
async function dirSize(target: string): Promise<number> {
  // `du` exits non-zero on unreadable subdirectories but still prints the total.
  const { stdout } = await execa('du', ['-sk', target], { reject: false });
  const kb = Number.parseInt(stdout.trim().split(/\s+/)[0] ?? '', 10);
  return Number.isFinite(kb) ? kb * 1024 : 0;
}

/**
 * Delete one DerivedData folder. Refuses anything that is not a direct child of
 * `root`, so the root itself and Xcode's shared caches can never be hit.
 */
export async function removeDerivedData(entry: DerivedDataEntry, root: string): Promise<void> {
  const target = path.resolve(entry.path);
  if (path.dirname(target) !== path.resolve(root) || !entry.name) {
    throw new Error(`Refusing to delete ${target}: not a project folder inside ${root}.`);
  }
  await fs.rm(target, { recursive: true, force: true });
}

export function formatBytes(bytes: number): string {
  if (bytes < 1024) return `${bytes} B`;
  const units = ['KB', 'MB', 'GB', 'TB'];
  let value = bytes / 1024;
  let unit = 0;
  while (value >= 1024 && unit < units.length - 1) {
    value /= 1024;
    unit++;
  }
  return `${value.toFixed(value < 10 ? 1 : 0)} ${units[unit]}`;
}

function isInside(parent: string, child: string): boolean {
  const rel = path.relative(parent, path.resolve(child));
  return rel !== '' && !rel.startsWith('..') && !path.isAbsolute(rel);
}

async function realpath(target: string): Promise<string> {
  try {
    return await fs.realpath(target);
  } catch {
    return path.resolve(target);
  }
}

async function exists(target: string): Promise<boolean> {
  try {
    await fs.access(target);
    return true;
  } catch {
    return false;
  }
}

File: src/derived_data.zig
const std = @import("std");
const Io = std.Io;
const disk = @import("disk.zig");
const exec = @import("exec.zig");

pub const Entry = struct {
    path: []const u8,
    name: []const u8,
    workspace_path: []const u8,
};

pub const Sized = struct {
    entry: Entry,
    size: u64,
};

pub const Error = error{RefusingToDelete} || std.mem.Allocator.Error;

pub fn root(
    gpa: std.mem.Allocator,
    io: Io,
    environ: *const std.process.Environ.Map,
) ![]const u8 {
    const home = environ.get("HOME") orelse return error.NoHomeDirectory;

    if (environ.get("LCC_DERIVED_DATA")) |raw| {
        const override = std.mem.trim(u8, raw, " \t");
        if (override.len > 0) {
            if (std.mem.eql(u8, override, "~")) return home;
            if (std.mem.startsWith(u8, override, "~/")) {
                return std.fs.path.join(gpa, &.{ home, override[2..] });
            }
            return gpa.dupe(u8, override);
        }
    }

    if (exec.capture(gpa, io, &.{
        "defaults", "read", "com.apple.dt.Xcode", "IDECustomDerivedDataLocation",
    }, null)) |custom| {
        if (custom.len > 0 and std.fs.path.isAbsolute(custom)) return custom;
    } else |_| {}

    return std.fs.path.join(gpa, &.{ home, "Library", "Developer", "Xcode", "DerivedData" });
}

pub fn list(gpa: std.mem.Allocator, io: Io, dir_path: []const u8) ![]Entry {
    var dir = Io.Dir.cwd().openDir(io, dir_path, .{ .iterate = true }) catch return &.{};
    defer dir.close(io);

    var entries: std.ArrayList(Entry) = .empty;
    var it = dir.iterate();
    while (it.next(io) catch null) |dirent| {
        if (dirent.kind != .directory or std.mem.startsWith(u8, dirent.name, ".")) continue;
        const name = try gpa.dupe(u8, dirent.name);
        const full = try std.fs.path.join(gpa, &.{ dir_path, name });
        const workspace = (try readWorkspacePath(gpa, io, full)) orelse continue;
        try entries.append(gpa, .{ .path = full, .name = name, .workspace_path = workspace });
    }
    return entries.toOwnedSlice(gpa);
}

fn readWorkspacePath(gpa: std.mem.Allocator, io: Io, dir_path: []const u8) !?[]const u8 {
    const plist = try std.fs.path.join(gpa, &.{ dir_path, "info.plist" });
    const raw = Io.Dir.cwd().readFileAlloc(io, plist, gpa, .limited(1 << 20)) catch return null;

    if (try matchWorkspacePath(gpa, raw)) |value| return value;

    const converted = exec.capture(gpa, io, &.{
        "plutil", "-convert", "xml1", "-o", "-", plist,
    }, null) catch return null;
    return matchWorkspacePath(gpa, converted);
}

fn matchWorkspacePath(gpa: std.mem.Allocator, xml: []const u8) !?[]const u8 {
    const key = "<key>WorkspacePath</key>";
    const key_at = std.mem.indexOf(u8, xml, key) orelse return null;
    const after = xml[key_at + key.len ..];

    const open_at = std.mem.indexOf(u8, after, "<string>") orelse return null;
    for (after[0..open_at]) |c| {
        if (!std.ascii.isWhitespace(c)) return null;
    }
    const value_start = open_at + "<string>".len;
    const close_at = std.mem.indexOfPos(u8, after, value_start, "</string>") orelse return null;

    const value = std.mem.trim(u8, after[value_start..close_at], " \t\r\n");
    if (value.len == 0) return null;
    return try decodeEntities(gpa, value);
}

fn decodeEntities(gpa: std.mem.Allocator, value: []const u8) ![]const u8 {
    const replacements = [_][2][]const u8{
        .{ "&lt;", "<" },
        .{ "&gt;", ">" },
        .{ "&quot;", "\"" },
        .{ "&apos;", "'" },
        .{ "&amp;", "&" },
    };
    var out: std.ArrayList(u8) = .empty;
    var i: usize = 0;
    outer: while (i < value.len) {
        if (value[i] == '&') {
            for (replacements) |pair| {
                if (std.mem.startsWith(u8, value[i..], pair[0])) {
                    try out.appendSlice(gpa, pair[1]);
                    i += pair[0].len;
                    continue :outer;
                }
            }
        }
        try out.append(gpa, value[i]);
        i += 1;
    }
    return out.toOwnedSlice(gpa);
}

pub fn forWorktree(
    gpa: std.mem.Allocator,
    io: Io,
    entries: []const Entry,
    worktree_path: []const u8,
) ![]Entry {
    const resolved = disk.realPath(gpa, io, worktree_path);

    var matched: std.ArrayList(Entry) = .empty;
    for (entries) |entry| {
        if (disk.isInside(gpa, resolved, entry.workspace_path)) try matched.append(gpa, entry);
    }
    return matched.toOwnedSlice(gpa);
}

pub fn orphans(gpa: std.mem.Allocator, io: Io, entries: []const Entry) ![]Entry {
    var dead: std.ArrayList(Entry) = .empty;
    for (entries) |entry| {
        Io.Dir.cwd().access(io, entry.workspace_path, .{}) catch {
            try dead.append(gpa, entry);
            continue;
        };
    }
    return dead.toOwnedSlice(gpa);
}

pub fn withSizes(gpa: std.mem.Allocator, io: Io, entries: []const Entry) ![]Sized {
    const paths = try gpa.alloc([]const u8, entries.len);
    for (entries, 0..) |entry, i| paths[i] = entry.path;
    const sizes = try disk.usage(gpa, io, paths);

    const sized = try gpa.alloc(Sized, entries.len);
    for (entries, 0..) |entry, i| sized[i] = .{ .entry = entry, .size = sizes[i] };
    return sized;
}

pub fn remove(gpa: std.mem.Allocator, io: Io, entry: Entry, dir_path: []const u8) !void {
    _ = gpa;
    if (entry.name.len == 0) return Error.RefusingToDelete;
    disk.removeChild(io, dir_path, entry.path) catch return Error.RefusingToDelete;
}

test "workspace path is pulled out of plist xml" {
    const gpa = std.testing.allocator;
    const xml =
        \\<plist><dict>
        \\<key>WorkspacePath</key>
        \\<string>/Users/me/Projects/App &amp; Co/App.xcodeproj</string>
        \\</dict></plist>
    ;
    const got = (try matchWorkspacePath(gpa, xml)).?;
    defer gpa.free(got);
    try std.testing.expectEqualStrings("/Users/me/Projects/App & Co/App.xcodeproj", got);
}

test "unrelated key is not mistaken for the workspace path" {
    const gpa = std.testing.allocator;
    const xml = "<key>Other</key><string>/nope</string>";
    try std.testing.expect((try matchWorkspacePath(gpa, xml)) == null);
}
